snekbox/snekio.py

```diff
@@ -0,0 +1,108 @@
+"""I/O Operations for sending / receiving files from the sandbox."""
+from __future__ import annotations
+
+import zlib
+from base64 import b64decode, b64encode
+from dataclasses import dataclass
+from pathlib import Path
+from typing import Generic, TypeVar
+
+RequestType = dict[str, str | bool | list[str | dict[str, str]]]
+
+T = TypeVar("T", str, bytes)
+
+
+def sizeof_fmt(num: int, suffix: str = "B") -> str:
+    """Return a human-readable file size."""
+    for unit in ("", "Ki", "Mi", "Gi", "Ti", "Pi", "Ei", "Zi"):
+        if abs(num) < 1024:
+            return f"{num:3.1f}{unit}{suffix}"
+        num /= 1024
+    return f"{num:.1f}Yi{suffix}"
+
+
+class AttachmentError(ValueError):
+    """Raised when an attachment is invalid."""
+
+
+class ParsingError(AttachmentError):
+    """Raised when an incoming file cannot be parsed."""
+
+
+class IllegalPathError(AttachmentError):
+    """Raised when an attachment has an illegal path."""
+
+
+@dataclass
+class FileAttachment(Generic[T]):
+    """A file attachment."""
+
+    name: str
+    content: T
+
+    @classmethod
+    def from_dict(cls, data: dict[str, str]) -> FileAttachment:
+        """Convert a dict to an attachment."""
+        name = data["name"]
+        path = Path(name)
+        parts = path.parts
+
+        if path.is_absolute() or set(parts[0]) & {"\\", "/"}:
+            raise IllegalPathError(f"File path '{name}' must be relative")
+
+        if any(set(part) == {"."} for part in parts):
+            raise IllegalPathError(f"File path '{name}' may not use traversal ('..')")
+
+        match data.get("content-encoding"):
+            case "base64":
+                content = b64decode(data["content"])
+            case None | "utf-8" | "":
+                content = data["content"].encode("utf-8")
+            case _:
+                raise ParsingError(f"Unknown content encoding '{data['content-encoding']}'")
```

Comment:
Let's not overcomplicate this. Similar to how the response always uses b64 for files, the request should also always use b64 for files.

Reply:
Made request accept base64 only, removed `content-encoding` parameter, in [4787a94](https://github.com/python-discord/snekbox/pull/159/commits/4787a9497be4f460e028031ddad4cd9d8388b84f)